text/XXXX-storage_command.md

```diff
@@ -0,0 +1,178 @@
+| net overhead | sum of blocks in each dag for _all_ commits in a history | 
+| gross overhead | how much disk space the total set of versions represents | sum of listed dataset sizes |
+
+* storage cost: how much space is this dataset taking up?
+* commit overhead: Which commit should I delete to get the most space back?
+* net overhead: how much space would it take to store this entire dataset?
+* gross overhead: how much space would it take up to write all of these versions to seperate files?
+
+The stats should present as much of this info as possible, scoped to the arguments provided.
+
+
+### purge: remove stored datasets without affecting history
+```
+$ qri storage purge --help
+Purge deletes dataset version data, freeing up disk space. Purge does not edit 
+history. To edit version history, use the remove command.
+
+Dataset versions take up disk space, and at some point need to be managed.
+Becase datasets and storage capacity vary dramatically the need for this
+
+You can also use purge to request a remote drop version data. This is often
+
+Usage:
+  qri purge [REMOTE] DATASET [flags]
+
+Examples:
+  # drop
+
+Flags:
+  -a, --all               purge all versions, overrides revisions
+  -r, --revisions         purge a number of versions from the latest commit
+  -t, --tail              purge stored versions starting from the earliest
+
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The purpose of stats are to answer a few primary questions. All of them deal with a _set of blocks_. 
+
+
+The foundation for all of this work should be a `dag.Info` & `dag.Manifest`. Info provides the block size information, and manifests provide the . `dag.Info` currently looks like this:
+
+```go
+// Info is os.FileInfo for dags: a struct that describes important
+// details about a graph. Info builds on a manifest.
+//
+// when being sent over the network, the contents of Info should be considered gossip,
+// as Info's are *not* deterministic. This has important implications
+// Info should contain application-specific info about a datset
+type Info struct {
+	// Info is built upon a manifest
+	Manifest *Manifest      `json:"manifest"`
+	Labels   map[string]int `json:"labels,omitempty"` // sections are lists of logical sub-DAGs by positions in the nodes list
+	Sizes    []uint64       `json:"sizes,omitempty"`  // sizes of nodes in bytes
+}
+```
+
+Here's the current Manifest data structure:
+```go
+type Manifest struct {
+  Links [][2]int `json:"links"` // links between nodes
+  Nodes []string `json:"nodes"` // list if CIDS contained in the DAG
+}
+```
+
+Manifests assume they describe one graph with a single root. To represent multiple graphs in the same manifest, we'll need to make one change:
+
+```go
+type Manifest struct {
+  Graphs [][][2]int `json:"links"` // links between nodes
```

Comment:
Does this have an impact on dsync?

Is one entire push process come off of one dag.Info, that iterates over the graphs in the manifest? Or is this just for storage checking purposes?

Also, where in the sync process does the 'rejection' happen based on the storage's retention strategy? If the entire dsync details arrives in the dag.Info (because we have the manifests of each version), then we can do it immediately.

Reply:
yes it most definitely does affect dsync. The reject can potentially happen in two places that  are part of the `dsync receive` lifecycle:

1. if the [dsync push preCheck hook](https://github.com/qri-io/dag/blob/master/dsync/dsync.go#L352) returns an error
2. if the [dsync finalCheck hook](https://github.com/qri-io/dag/blob/5cd938b03caf29397ad78d79b4f4830dd48ae54f/dsync/dsync.go#L415-L418) returns an error.

qri implements these dsync hooks to do storage checks and supply a higher-order set of hooks. Qri always checks `acceptSizeMax` in the `dsync.FinalCheck` hook.

> Also, where in the sync process does the 'rejection' happen based on the storage's retention strategy? If the entire dsync details arrives in the dag.Info (because we have the manifests of each version), then we can do it immediately.

Totally agreed. This is an optimization we should be shipping for dsync.